// @mui/x-charts-pro/node/context/ZoomProvider/defaultizeZoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultizeZoom = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _internals = require("@mui/x-charts/internals");
const defaultZoomOptions = {
  minStart: 0,
  maxEnd: 100,
  step: 5,
  minSpan: 10,
  maxSpan: 100,
  panning: true,
  filterMode: 'keep'
};
const defaultizeZoom = (axis, axisDirection) => {
  if (!axis) {
    return [];
  }
  const defaultized = axis.map(v => {
    if (!v.zoom) {
      return undefined;
    }
    if (v.zoom === true) {
      return (0, _extends2.default)({
        axisId: v.id,
        axisDirection
      }, defaultZoomOptions);
    }
    return (0, _extends2.default)({
      axisId: v.id,
      axisDirection
    }, defaultZoomOptions, v.zoom);
  }).filter(_internals.isDefined);
  return defaultized;
};
exports.defaultizeZoom = defaultizeZoom;
