// @mui/x-charts-pro/Heatmap/HeatmapPlot.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useXScale, useYScale, useZColorScale } from '@mui/x-charts/hooks';
import { useHeatmapSeries } from "../hooks/useSeries.js";
import { HeatmapItem } from "./HeatmapItem.js";
import { jsx as _jsx } from "react/jsx-runtime";
function HeatmapPlot(props) {
  const xScale = useXScale();
  const yScale = useYScale();
  const colorScale = useZColorScale();
  const series = useHeatmapSeries();
  const xDomain = xScale.domain();
  const yDomain = yScale.domain();
  if (!series || series.seriesOrder.length === 0) {
    return null;
  }
  const seriesToDisplay = series.series[series.seriesOrder[0]];
  return /*#__PURE__*/_jsx("g", {
    children: seriesToDisplay.data.map(([xIndex, yIndex, value], dataIndex) => {
      const x = xScale(xDomain[xIndex]);
      const y = yScale(yDomain[yIndex]);
      const color = colorScale?.(value);
      if (x === undefined || y === undefined || !color) {
        return null;
      }
      return /*#__PURE__*/_jsx(HeatmapItem, {
        width: xScale.bandwidth(),
        height: yScale.bandwidth(),
        x: x,
        y: y,
        color: color,
        dataIndex: dataIndex,
        seriesId: series.seriesOrder[0],
        value: value,
        slots: props.slots,
        slotProps: props.slotProps
      }, `${xIndex}_${yIndex}`);
    })
  });
}
process.env.NODE_ENV !== "production" ? HeatmapPlot.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: PropTypes.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: PropTypes.object
} : void 0;
export { HeatmapPlot };
